pseudocode-defns/pseudocode_defns.tex

```diff
@@ -0,0 +1,490 @@
+
+\questionr{Should we have a general definition for ``floats'' (and ``integers''?), or is it sufficiently understood what a float is in general?}
+
+\section{Domains}
+
+$\sfi$ A \emph{data domain} is a representation of the set of values on which a metric or function can operate. For example, if a function accepts inputs from the domain \texttt{IntervalDomain(1:u32,17:u32)}, this means that the function can take any input value \texttt{v} of type \texttt{u32} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\questionr{Should anything else be included in this introduction to ``Domains''?}
+
+\begin{definition}[\texttt{AllDomain}]$\sfi$
+\texttt{AllDomain(T)} is the data domain consisting of all values of type $\T$. 
+\end{definition}
+
+For example, \texttt{AllDomain(u32)} is the domain of all values of type \texttt{u32}.
+
+\begin{definition}[\texttt{IntervalDomain}] \sfi
+For a type \texttt{T} that has a total ordering (\texttt{T} has trait \texttt{TotalOrd}) and for values \texttt{L <= U} of type \texttt{T}, \texttt{IntervalDomain(L:T, U:T)} is the domain that contains all values \texttt{v} of type \texttt{T} such that \texttt{L <= v} and \texttt{v <= U}.
+    
+Note that, because both \texttt{L} and \texttt{U} are of type \texttt{T}, there is no need to explicitly pass \texttt{T}; the type \texttt{T} can be inferred. \texttt{IntervalDomain} is defined on any type that implements the trait \texttt{TotalOrd}.\footnote{As of June 28, the OpenDP library requires the weaker condition of partial ordering (implements \texttt{PartialOrd}) instead.}
+\end{definition}
+
+For example, \texttt{IntervalDomain(1:u32, 17:u32)} corresponds to a domain that contains all the \texttt{u32} values \texttt{v} such that \texttt{1 <= v} and \texttt{v <= 17}.
+
+\begin{definition}[\texttt{SizedDomain}] $\sfi$
+    \texttt{SizedDomain(D)} is a domain that specifies the length of the enclosed domain \texttt{D}. The length of elements of the enclosed domain is specified with an element of type \texttt{usize}.
+\end{definition}
+
+For example, \texttt{SizedDomain(VectorDomain(AllDomain(u32), n)} corresponds to the domain of all vectors of length $\texttt{n}$ and elements of type \texttt{u32}.
+
+\begin{definition}[\texttt{VectorDomain}]
+\texttt{VectorDomain(D)} is the domain of all vectors of elements drawn from domain \texttt{D}.
+\end{definition}
+
+For example, \texttt{VectorDomain(AllDomain(u32))::Carrier} is \texttt{Vec(u32)}.
+
+\horizline
+
+\subsection{Notes, todos, questions}
+\todonei{Add clampable domain (\texttt{ClampableDomain}) -- waiting until \texttt{TotalOrd} is fully implemented in the OpenDP library.}
+
+\section{Traits}
+% any traits should be defined here
+
+\begin{definition}[\texttt{Bounded}]
+    A type $\T$ has trait \texttt{Bounded} $\iffText$ $\T$ has some upper bound and some lower bound (some smallest possible value and some largest possible value).
```

Comment:
So far the types described in the Types section are only the atomic types.   Is this referring to domains?  The doc doesn't say anywhere that domains are implemented as types.

Reply:
This "Bounded" trait applies to types (not domains). An example of a type which implements Bounded is u32: the smallest value is 0, and the largest value is 2^32-1. See [](https://docs.rs/num/0.4.0/num/traits/trait.Bounded.html) for more information on the "Bounded" trait.

Reply:
Actually, don't pay attention to the reply above yet. I'll work on learning more about "Bounded".

Reply:
OK, so my initial comment above is (I think) correct; domains are not the only types with the `Bounded` trait. We were definitely unclear, though, on what sorts of types domains have. I am working on fixing that now. 

As a quick example, `IntervalDomain(1:u32,17:u32)` has type `IntervalDomain[u32]`. That type `IntervalDomain[u32]` would, in turn, have the "`Bounded`" trait. The fact that "`IntervalDomain(1:u32,17:u32)` has type `IntervalDomain[u32]`" cannot be found in the current version of the document, but it will be added shortly.

Reply:
Thanks. Can you point me to an example of where the Bounded trait is useful?

Reply:
The purpose of bounded was to define the max value that an int may take on. In the count transformation, if the cast from usize to T failed, then the max value would be substituted. This behaviour is now encoded within the ExactIntCast trait.